client/asset/xmr/toolsdl/download.go

```diff
@@ -0,0 +1,869 @@
+	for i, zip := range d.hzips {
+		if !zip.valid {
+			continue
+		}
+		// similar os's
+		if !(d.machine.os == zip.os) &&
+			!(d.machine.os == "darwin" && zip.os == "mac") && !(d.machine.os == "windows" && zip.os == "win") {
+			continue
+		}
+		// similar arch's
+		if !(d.machine.arch == zip.arch) &&
+			!(d.machine.arch == "amd64" && zip.arch == "x64") && !(d.machine.arch == "arm64" && zip.arch == "armv8") {
+			continue
+		}
+		d.Log.Tracef("Chose zip for machine: %s %s", d.machine.os, d.machine.arch)
+		d.hzips[i].selected = true // maybe remove
+		d.selectedZip = &d.hzips[i]
+		return nil
+	}
+	return fmt.Errorf("no suitable zip found for os: %s, arch: %s", d.machine.os, d.machine.arch)
+}
+
+// ================================== Run ======================================
+// Run downloads a new set of dex monero tools of the latest version if it does
+// not already exist on the filesystem. If there are no dex monero tools already
+// and download also fails then we download of one of our JSON version sets.
+// =============================================================================
+func (d *Download) Run() (string, error) {
+	noLocal := false
+	noZip := false
+
+	d.machine = getMachine()
+
+	// working dir
+	tempDir, err := os.MkdirTemp("", "share-mtools-") // unix: /tmp/share-mtools-1234567
+	if err != nil {
+		noZip = true
+	}
+	d.tempDir = tempDir
+
+	// what do we have locally
+	hasPath, bestLocalToolsDir, bestLocalDirName, err := d.GetBestCurrentLocalToolsDir()
+	if err != nil || !hasPath {
+		noLocal = true
+	}
+	localVer, err := newMoneroVersionFromDir(bestLocalDirName)
+	if err != nil {
+		noLocal = true
+	}
+
+	// what do we have remotely
+	d.downloadRemoteCanonicalHashesFile()
+	hzip := d.getSelectedZip()
+	if hzip == nil {
+		noZip = true
+	}
+
+	// 1. only hard coded
+	if noLocal && noZip {
+		return d.runMavDownload()
+	}
+
+	// 2. canon first, hard coded second
+	if noLocal && !noZip {
+		toolsDir, err := d.runDownload()
+		if err != nil {
+			return d.runMavDownload()
+		}
+		return toolsDir, nil
+	}
+
+	// 3. local unless hard coded higher version
+	if !noLocal && noZip {
+		hcmvs, _ := getMoneroMAVersionSet()
+		var hcVer = moneroVersionZeroV0()
+		for _, hcmv := range hcmvs {
+			// hard coded is greater than local version.
+			// This cannot happen yet from the hard coded json
+			if hcmv.compare(localVer) > 0 {
+				hcVer = hcmv
+			}
+		}
+		if hcVer.compare(localVer) <= 0 { // local as good or better than hard coded
+			return bestLocalToolsDir, nil
+		}
+		mavToolsDir, err := d.runMavDownload()
+		if err != nil {
+			return bestLocalToolsDir, nil
+		}
+		return mavToolsDir, nil
+	}
+
+	// 4. got a best local, a valid hashedZip set.
+	if !noLocal && !noZip {
+		remVer := hzip.version
+		if remVer.compare(localVer) > 0 {
+			toolsDir, err := d.runDownload()
+			if err != nil {
+				return bestLocalToolsDir, nil
+			}
+			return toolsDir, nil
+		}
+		return bestLocalToolsDir, nil
+	}
+
+	return "", nil
```

Comment:
Is it not expected to get here? If not maybe return some error and comment here that we should never get here. These four ifs could also be a switch may be easier to see the paths. They wouldnt have one variable they are comparing to just:

```
switch {
    case a && b:
    case !a && b:
....
}


```